### PR TITLE
Add leaderboard edit section

### PR DESCRIPTION
**Pull request by gautamjajoo**
Add leaderboard edit section on settings tab where hosts can change:

- [x] Leaderboard Visiblity
- [x] Leaderboard Schema

**NOTE:**  
1. We will add leaderboard schema edit feature once we get the endpoint.
2. `is_leaderboard_public` parameter needs to changed, but it can only be changed once the `phase` PR is merged.

### CODE REVIEW

**Comment by gautamjajoo** [comment]
Thank You @Kajol-Kumari for the fix. Here is the updated screenshot:
Screenshot from 2021-07-16 22-14-06

**Comment by gautamjajoo** [comment]
So, we were getting two values of the variables which was being detected by `ngOnChanges`. The issue was that `ngOnChanges` detects change in the `@Input` variable, and when we were updating any variable inside it, it was called twice for both the `@Input` variable. So, we have added a check if the call was for the specific variable (`selectedPhase` or `selectedPhaseSplit`) and therefore this fixes our issue.

Also, I have removed `ngModel` from selectPhase component as it is not being used there.

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html [hunk 1]
```diff
@@ -108,35 +108,71 @@ <h5 class="fw-light">Challenge Settings</h5>
...
+            #phasesplitselect
+          >
+          </app-selectphase>
+          </div>
+        <div class="row margin-bottom-cancel phase-card" *ngIf="selectedPhaseSplit">
+          <div class="col s12">
+            <span class="fs-16 fw-light">
+              <strong class="text-light-black">Leaderboard Schema </strong>
+            </span>
+            <a class="pointer fs-16" (click)="editPhaseDetails()">
```
Comment (Kajol-Kumari) [inline frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html:166]: [Question] Why are we calling `editPhaseDetails()` function here? As the backend API is not ready so we can add an empty function for now having a `TODO` comment in the controller

**Review by Kajol-Kumari — APPROVED** [review]
@gautamjajoo minor change required, rest LGTM 👍

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html (hunk 1 above)
Comment (gautamjajoo) [reply frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html]: Done the changes!

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts [hunk 2]
```diff
@@ -121,14 +121,57 @@ export class ChallengesettingsComponent implements OnInit, OnDestroy {
...
+   */
+  leaderboardVisibility = {
+    state: 'Private',
+    icon: 'fa fa-toggle-off',
+  };
+
+  /**
+   * If leaderboard of phase split is public
+   */
+   isPhaseSplitLeaderboardPublic: number = 1;
```
Comment (Ram81) [inline frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts:155]: this should be boolean

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts (hunk 2 above)
Comment (gautamjajoo) [reply frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts:155]: Actually, phase split visibility is defined by number(`1` or `3`)

**Comment by gautamjajoo** [comment]
@Ram81 @Kajol-Kumari Done all the required changes. Now we don't fetch the phases/phase splits again. Also, the selected phase/phase split is also displayed with all the changes visible on the other components.

**Review by Kajol-Kumari — CHANGES_REQUESTED** [review]
The `private` tag isn't updating correctly after we change the phase/phase-split/submission visibility. Please fix it for both phase edit and leaderboard edit section -
Screenshot 2021-07-20 at 8 47 56 PM

**Comment by gautamjajoo** [comment]
@Kajol-Kumari I believe we cannot change that because how would we update the already selected phase?

**Comment by gautamjajoo** [comment]
@Kajol-Kumari Found a solution for the fix. Updated the PR as well.

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts [hunk 3]
```diff
@@ -212,6 +280,104 @@ export class ChallengesettingsComponent implements OnInit, OnDestroy {
...
+      title: 'Edit Leaderboard Schema',
+      label: 'leaderboard_schema',
+      isEditorRequired: true,
+      confirm: 'Submit',
+      deny: 'Cancel',
+      confirmCallback: SELF.apiCall,
+    };
+    SELF.globalService.showModal(PARAMS);
+
+    }
```
Comment (Ram81) [inline frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts:379]: Let's remove this option for now. We can add this when I have the API ready. Probably next week, I won't have bandwidth to work on this before that

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts [hunk 4]
```diff
@@ -212,6 +280,104 @@ export class ChallengesettingsComponent implements OnInit, OnDestroy {
...
+      .patchFileUrl(
+        SELF.endpointsService.particularChallengePhaseSplitUrl(SELF.selectedPhaseSplit['id']),
+        BODY
+      )
+        .subscribe(
+          (data) => {
+            SELF.selectedPhaseSplit['visibility'] = data.visibility;
+            SELF.challengeService.changePhaseSplitSelected(true);
+            if (visibility == 3) {
+              SELF.selectedPhaseSplit['showPrivate'] = false;
```
Comment (Ram81) [inline frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts:331]: Fix the indentation

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts [hunk 5]
```diff
@@ -212,6 +280,104 @@ export class ChallengesettingsComponent implements OnInit, OnDestroy {
...
+       SELF.selectedPhaseSplit = phaseSplit;
+       SELF.isPhaseSplitLeaderboardPublic = SELF.selectedPhaseSplit['visibility'];
+       if(SELF.isPhaseSplitLeaderboardPublic == 3) {
+         SELF.leaderboardVisibility.state = 'Public';
+         SELF.leaderboardVisibility.icon = 'fa fa-toggle-on green-text';
+       }
+       else {
+        SELF.leaderboardVisibility.state = 'Private';
+        SELF.leaderboardVisibility.icon = 'fa fa fa-toggle-off grey-text text-darken-1';
+       }
```
Comment (Ram81) [inline frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts:298]: Indentation spacing looks inconsistent. It should be two spaces

**Review by Ram81 — CHANGES_REQUESTED** [review]

**Comment by Ram81** [comment]
@gautamjajoo let's remove the `Leaderboard Schema` option from the edit page for now. We can add it back when API is ready

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html [hunk 6]
```diff
@@ -108,35 +108,71 @@ <h5 class="fw-light">Challenge Settings</h5>
...
+          </a>
+        </div>
+       </div>
+    </div>
+  </div>
+
+    <!-- Leaderboard Edit Section-->
+    <div class="row settings-section">
+      <div class="col-lg-3 col-md-4 col-sm-3 col-xs-3">
+        <span class="fw-light fs-18">Edit leaderboard details:</span>
```
[the remark this answers is not in the document]
Comment (gautamjajoo) [reply frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html:146]: @Ram81 To make things consistent should I remove `:` from heading of every section?

[the remark this answers is not in the document]
Comment (gautamjajoo) [reply frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html:146]: @Ram81 To make things consistent should I remove `:` from heading of every section?

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts (hunk 3 above)
Comment (gautamjajoo) [reply frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.ts:379]: Sure, I have removed it

**Comment by gautamjajoo** [comment]
@Ram81 Done the required changes.

**Review by Kajol-Kumari — APPROVED** [review]
LGTM 👍

**Review by Ram81 — APPROVED** [review]
LGTM

File: frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ram81) [reply frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html:146]: Yes, we should remove `:` from all headings

[the remark this answers is not in the document]
Comment (gautamjajoo) [reply frontend_v2/src/app/components/challenge/challengesettings/challengesettings.component.html]: Done!

**Pull request merged**